lib/src/rules/avoid_unstable_final_fields.dart

```diff
@@ -0,0 +1,685 @@
+  B1(super.i);
+}
+
+class B2 implements A {
+  int i; // LINT.
+  B2(this.i);
+}
+```
+
+**GOOD:**
+```dart
+class C {
+  final int i;
+  C(this.i);
+}
+
+class D1 implements C {
+  late final int i = someExpression; // OK.
+}
+
+class D2 extends C {
+  int get i => super.i + 1; // OK.
+  D2(super.i);
+}
+```
+
+''';
+
+class AvoidUnstableFinalFields extends LintRule {
+  AvoidUnstableFinalFields()
+      : super(
+            name: 'avoid_unstable_final_fields',
+            description: _desc,
+            details: _details,
+            group: Group.errors);
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    var visitor = _Visitor(this, context);
+    registry.addFieldDeclaration(this, visitor);
+    registry.addMethodDeclaration(this, visitor);
+  }
+}
+
+bool _isLocallyStable(Element element) {
+  if (element is PropertyAccessorElement &&
+      element.isSynthetic &&
+      element.correspondingSetter == null) {
+    // This is a final, non-local variable, and they are stable.
+    var metadata = element.variable.metadata;
+    if (metadata.isNotEmpty) {
+      for (var elementAnnotation in metadata) {
+        var metadataElement = elementAnnotation.element;
+        if (metadataElement is ConstructorElement) {
+          var metadataOwner = metadataElement.declaration.enclosingElement;
+          if (metadataOwner.isDartCoreObject) {
+            // A declaration with `@Object()` is not considered stable.
+            return false;
+          }
+        }
+      }
+    }
+    return true;
+  } else if (element is FunctionElement) {
+    // A tear-off of a top-level function or static method is stable,
+    // local functions and function literals are not.
+    return element.isStatic;
+  } else if (element is ClassElement) {
+    // A reified type of a class is stable.
+    return true;
+  } else if (element is MethodElement) {
+    // An instance method tear-off is never stable.
+    return false;
+  }
+  // TODO(eernst): Any cases still missing?
+  return false;
+}
+
+abstract class _AbstractVisitor extends ThrowingAstVisitor<void> {
+  final LintRule rule;
+  final LinterContext context;
+
+  // Will be true initially when a getter body is traversed. Will be made
+  // true if the getter body turns out to be unstable. Is checked after the
+  // traversal of the body, ot emit a lint if it is still false at that time.
+  bool isStable = true;
+
+  // Each [MethodDeclaration] which is causing the lint to be reported is added
+  // to this list, which is then used to create context messages.
+  var causes = <PropertyAccessorElement>{};
+
+  // Initialized in `visitMethodDeclaration` if a lint might be emitted.
+  // It is then guaranteed that `declaration.isGetter` is true.
+  late final MethodDeclaration declaration;
+
+  _AbstractVisitor(this.rule, this.context);
+
+  bool _inheritsStability(
+      ClassElement classElement, Name name, LinterContext context) {
```

Comment:
You probably don't need to pass the `LinterContext`, here and below, given that it's visible as an instance field.